### DISCUSSION TITLE
Build npm packages dynamically / on-demand?

### DISCUSSION THREAD

**Discussion by borekb** [verdaccio/verdaccio#2127] [Ideas]
I'm new to Verdaccio so maybe I'm wrong but I think you have to _publish_ to it, or set it up as a proxy to another registry. In other words, it contains static tarballs. I'm wondering if there is (or could be) an on-demand approach, useful for local testing. 

Let's say I have a set of Yarn workspaces and many of the packages are not published to npm yet, or at least not in their latest versions. I then have a test app that is supposed to test the packages. I imagine this test app would contain something like this in its `.yarnrc.yml`:

```yaml
npmScopes:
  my-company:
    npmRegistryServer: //localhost:1234  # Verdaccio
```

and when it gets hit with requests for packages like `@my-company/example@1.2.3`, it would **build** the tarball dynamically (and probably cache it).

Is there such dynamic mode currently? Could it be a feature in the future if not?

**Comment by jachstet-sea** [comment]
I would say that this is more a task for a `continuous integration` (CI) server instead of verdaccio. Building and testing packages automatically is a very good thing. However, there are so many things that differ from package to package so that a lot of factors would have to be taken into account. And with existing CI servers and solutions, there is already software specialized in that. No one stops you from configuring your CI server to automatically publish to your NPM server after all tests have passed.

**Reply by borekb** [reply]
CI feels to late. I'm making a change locally to my source code, I want to test it even before I create a commit, let alone push to some remote that triggers a CI process.

But I understand this is not what Verdaccio was build for – fair enough.

**Reply by jachstet-sea** [reply]
Okay, now I see what you have in mind. However, if you want this to work "without committing the changes" to your VCS, this means you have to run verdaccio on your developer's machine?

Beware that I'm only sharing my personal opinion, it's not the official opinion of `the Verdaccio project` or their maintainers.

**Reply by borekb** [reply]
Yes, locally, as in its most basic usage. I don't want to `npm publish` into it, I want Verdaccio to watch incoming requests for my specific packages (`@my-company/package-foo`) and when that happens, build the tarball on-demand.
